src/org/sosy_lab/common/collect/OrderStatisticSet.java

```diff
@@ -0,0 +1,113 @@
+/*
+ *  SoSy-Lab Common is a library of useful utilities.
+ *  This file is part of SoSy-Lab Common.
+ *
+ *  Copyright (C) 2007-2015  Dirk Beyer
+ *  All rights reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.sosy_lab.common.collect;
+
+import java.util.SortedSet;
+
+/**
+ * A {@link SortedSet} that allows two additional operations: receiving (and deleting) an element by
+ * its <i>rank</i>, and getting the rank of an element.
+ *
+ * <p>Implementations should adhere to all contracts of the <code>SortedSet</code> interface.
+ *
+ * <p>Implementing classes should provide two means for comparing elements:
+ *
+ * <ol>
+ *   <li>Using the natural ordering of the elements. In this case, all elements of the set have to
+ *       implement the {@link Comparable} interface.
+ *   <li>Using a {@link java.util.Comparator Comparator} to create an order over the elements of the
+ *       set.
+ * </ol>
+ *
+ * In both cases, the used compare-method should be consistent with <code>equals</code>, i.e.,
+ * <code>compare(a, b) == 0  =&gt;  a.equals(b)</code>, so that the contract provided by {@link
+ * java.util.Set Set} is fulfilled. Implementations of this interface should still work if the used
+ * compare-method is not consistent with <code>equals</code>.
+ *
+ * <p>If the used compare-method is not consistent with <code>equals</code>, unequal (i.e.,
+ * different) elements that are equal according to the order can exist in a set. In this case,
+ * elements that are added first will have a higher rank than elements of the same order that are
+ * added later.
```

Comment:
Are you sure you want to guarantee a specific behavior if `compareTo` and `equal` are inconsistent? Why not simply require that they are consistent?

Also, this is a contradiction to the paragraph above, where it simply says that implementations "should" work (but are not required to) if the mthods are inconsistent.

Reply:
Requiring that the used `compare`-method (not just `compareTo`!) is consistent with `equals`  is a strong constraint on the possible uses of the interface.

Whenever you want to use some comparison criterion that is not consistent with `equals`,
you will have to work around this by maintaining collections of objects for each 'comparison value', manually.

Through this definition in the interface, we would allow implementing sets to do this for the user.

> Also, this is a contradiction to the paragraph above, where it simply says that implementations "should" work (but are not required to) if the mthods are inconsistent.

Thanks!

Reply:
`SortedSet` also requires consistency of `equals` and `compareTo`, at least for fulfilling the `Set` contract. We should use the same guarantees and requirements.

`SkipList` doesn't call `equals` at all, right? Then it will work anyway just like users expect it.

Reply:
Yes, the `SortedSet` interface states:
> […] if the sorted set is to correctly implement the Set interface

So that's equal to our condition, isn't it?

>SkipList doesn't call equals at all, right? Then it will work anyway just like users expect it.

Yes it does. All operations that are specific to a given object (e.g., `remove`, `contains`) use the equals-method to ensure that the objects are not only equal by the given Comparator, but by their own equals-method.

Reply:
> All operations that are specific to a given object (e.g., remove, contains) use the equals-method to ensure that the objects are not only equal by the given Comparator, but by their own equals-method.

This should not be done. Just use only the given order, and document that if `equals` and the given order is inconsistent, then the `Set` contract is not fulfilled. This is the same as all other `SortedSet` implementations I know do.

Reply:
Oh ok, I can do that.